memstore/list/memory_pool.go

```diff
@@ -0,0 +1,171 @@
+	// Return the actual memory address given offset.
+	Interpret(offset uintptr) uintptr
+	// Release the underlying memory.
+	Destroy()
+}
+
+// slabMemoryPool implements the HighLevelMemoryPool using slab algorithm.
+type slabMemoryPool struct {
+	slabAllocator    *Arena
+	nativeMemoryPool NativeMemoryPool
+}
+
+// NewHighLevelMemoryPool returns a default implementation of HighLevelMemoryPool.
+func NewHighLevelMemoryPool(reporter HostMemoryReporter) HighLevelMemoryPool {
+	nativeMP := NewNativeMemoryPool(reporter)
+	slabAllocator :=
+		NewArena(defaultStartSlabAllocatorChunkSize, defaultSlabSize, defaultSlabGrowthFactor, nativeMP)
+	return slabMemoryPool{
+		slabAllocator:    slabAllocator,
+		nativeMemoryPool: nativeMP,
+	}
+}
+
+// Allocate is the implementation of Allocate in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Allocate(size int) [2]uintptr {
+	if size == 0 {
+		return [2]uintptr{0, 0}
+	}
+	return mp.slabAllocator.Alloc(size)
+}
+
+// Reallocate is the implementation of Reallocate in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Reallocate(oldBuf [2]uintptr, oldSize int, newSize int) [2]uintptr {
+	if oldSize == newSize {
+		return oldBuf
+	}
+
+	if oldSize != 0 {
+		mp.Free(oldBuf)
+	}
+
+	return mp.Allocate(newSize)
+}
+
+// Free is the implementation of Free in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Free(buf [2]uintptr) {
+	if !mp.slabAllocator.DecRef(buf) {
+		utils.GetLogger().Panic("buf should be freed but not")
+	}
+}
+
+// Interpret is the implementation of Interpret in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Interpret(offset uintptr) uintptr {
+	return offset + mp.nativeMemoryPool.GetBaseAddr()
+}
+
+// Destroy is the implementation of Free in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Destroy() {
+	mp.nativeMemoryPool.Destroy()
+	mp.nativeMemoryPool = nil
+}
+
+// NativeMemoryPool is the interface to manage system memory to support high level memory pool
+// allocation requests. All the pointer/address returned by this memory pool is relative to the
+// base address fetched via GetBaseAddr.
+type NativeMemoryPool interface {
+	// Malloc returns a byte slice to caller, will allocate more memoryOffset if no enough space.
+	// returned addresses are not aligned.
+	Malloc(size int) uintptr
+	// Destroy frees the memory managed by this memory pool. After destroy, any malloc call's
+	// behaviour will be undefined.
+	Destroy()
+	// GetBaseAddr returns the base address managed by this pool.
+	GetBaseAddr() uintptr
+}
+
+// singleChunkNativeMemoryPool manages the system memory as a single chunk of continuous memory address.
+// When this memory pool can no longer satisfy allocation requests, it will allocate another memory chunk
+// big enough to hold the old memory content while satisfying the new requests. Then it will make a memcpy
+// to move the old data into the new location. A small performance penalty will be paid during this period.
```

Comment:
this is prefered over a linked list implementation because we need continuous memory space right?

Reply:
yes